### ISSUE TITLE
Parse from `R: Read` instead of byte slice

### ISSUE THREAD

**Issue by Vurich** [yurydelendik/wasmparser.rs#90]
Even though this parser is streaming, it's currently not possible to actually have it operate on a stream. The API requires a `&[u8]`. This isn't really necessary though, most of the API could remain identical while supporting any input type `R where R: Read`. I tried to implement this myself but this codebase is so huge and has so much duplicated code that I gave up, since it's not immediately necessary for my use-case.

**Comment by yurydelendik** [comment]
Yes, that is the change that needs to be made. I need to investigate the performance impact first (and was hoping #8 is addressed first)

**Comment by Vurich** [comment]
It's possible to use specialisation to allow byte slices/vectors to continue to slice into the input for references (such as names) while other `Read` impls collect into a vector, which should eliminate the only possible performance impact.

**Comment by Robbepop** [comment]
I wonder what it would take to actually implement this initially for this crate. It seems to be super important to make this crate actually support streaming parsing applications that are out there in the wild. ;)

I guess for things like the `BrTable` that needs double-parsing this change to `T: Read` might be tricky but in general it should be implementable and shouldn't have significant performance impacts if we buffer our reads from the sources correctly.

@yurydelendik What do you think about this?
